File: C#/BluffinPokerServer/ServerLobby.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Net.Sockets;
using PokerWorld.Game;
using System.Threading;
using PokerProtocol.Commands.Lobby;
using PokerProtocol;
using PokerProtocol.Commands.Lobby.Training;
using PokerProtocol.Commands.Lobby.Career;
using System.Net;
using EricUtility;

namespace BluffinPokerServer
{
    public class ServerLobby
    {
        private readonly int m_NoPort;
        private readonly TcpListener m_SocketServer;

        private readonly List<string> m_UsedNames = new List<string>();
        private readonly Dictionary<int, PokerGame> m_Games = new Dictionary<int, PokerGame>();

        private int m_LastUsedID = 0;

        public PokerGame GetGame(int id)
        {
            return m_Games[id];
        }

        public ServerLobby(int port)
        {
            m_NoPort = port;
            m_SocketServer = new TcpListener(IPAddress.Any,port);
        }

        public bool NameUsed(string name)
        {
            foreach (string s in m_UsedNames)
                if (s.Equals(name, StringComparison.InvariantCultureIgnoreCase))
                    return true;
            return false;
        }

        public void AddName(string name)
        {
            m_UsedNames.Add(name);
        }

        public void RemoveName(string name)
        {
            m_UsedNames.Remove(name);
        }

        public void Start()
        {
            m_SocketServer.Start();
            new Thread(new ThreadStart(Run)).Start();
        }

        private void Run()
        {
            while (true)
            {
                try
                {
                    TcpClient client = m_SocketServer.AcceptTcpClient();
                    ServerClientLobby lobby = new ServerClientLobby(client, this);
                    lobby.Start();
                }
                catch (Exception e)
                {
                    LogManager.Log(LogLevel.Error, "ServerLobby.Run", e.StackTrace);
                }
            }
        }

        public int CreateTrainingTable(CreateTrainingTableCommand c)
        {
            ListTrainingTables();
            ListCareerTables();
            m_LastUsedID++;
            while (m_Games.ContainsKey(m_LastUsedID))
                m_LastUsedID++;
            PokerGameTraining game = new PokerGameTraining(new TableInfoTraining(c.TableName, c.BigBlind, c.MaxPlayers, c.Limit, c.StartingMoney), c.WaitingTimeAfterPlayerAction, c.WaitingTimeAfterBoardDealed, c.WaitingTimeAfterPotWon);
            m_Games.Add(m_LastUsedID, game);
            game.Start();
            return m_LastUsedID;
        }

        public int CreateCareerTable(CreateCareerTableCommand c)
        {
            ListTrainingTables();
            ListCareerTables();
            m_LastUsedID++;
            while (m_Games.ContainsKey(m_LastUsedID))
                m_LastUsedID++;
            PokerGameCareer game = new PokerGameCareer(new TableInfoCareer(c.TableName, c.BigBlind, c.MaxPlayers, c.Limit), c.WaitingTimeAfterPlayerAction, c.WaitingTimeAfterBoardDealed, c.WaitingTimeAfterPotWon);
            m_Games.Add(m_LastUsedID, game);
            game.Start();
            return m_LastUsedID;
        }

        public List<TupleTableInfoTraining> ListTrainingTables()
        {
            List<TupleTableInfoTraining> tables = new List<TupleTableInfoTraining>();
            List<int> tablesToRemove = new List<int>();

            foreach (KeyValuePair<int, PokerGame> kvp in m_Games)
            {
                PokerGame game = kvp.Value;
                int noPort = kvp.Key;
                if (game.IsRunning)
                {
                    TableInfo t = game.Table;
                    if( t is TableInfoTraining )
                        tables.Add(new TupleTableInfoTraining(noPort, t.Name, t.BigBlindAmnt, t.Players.Count, t.NbMaxSeats, t.BetLimit, PossibleActionType.None));
                }
                else
                    tablesToRemove.Add(noPort);
            }
            foreach (int i in tablesToRemove)
                m_Games.Remove(i);
            return tables;
        }

        public List<TupleTableInfoCareer> ListCareerTables()
        {
            List<TupleTableInfoCareer> tables = new List<TupleTableInfoCareer>();
            List<int> tablesToRemove = new List<int>();

            foreach (KeyValuePair<int, PokerGame> kvp in m_Games)
            {
                PokerGame game = kvp.Value;
                int noPort = kvp.Key;
                if (game.IsRunning)
                {
                    TableInfo t = game.Table;
                    if (t is TableInfoCareer)
                        tables.Add(new TupleTableInfoCareer(noPort, t.Name, t.BigBlindAmnt, t.Players.Count, t.NbMaxSeats, t.BetLimit, PossibleActionType.None));
                }
                else
                    tablesToRemove.Add(noPort);
            }
            foreach (int i in tablesToRemove)
                m_Games.Remove(i);
            return tables;
        }
    }
}

File: trunk/JAVA/BluffinServer/src/bluffinmuffin/server/ServerLobby.java
package bluffinmuffin.server;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import bluffinmuffin.poker.PokerGame;
import bluffinmuffin.poker.PokerGameCareer;
import bluffinmuffin.poker.PokerGameTraining;
import bluffinmuffin.poker.entities.TableInfo;
import bluffinmuffin.poker.entities.TableInfoCareer;
import bluffinmuffin.poker.entities.TableInfoTraining;
import bluffinmuffin.protocol.PossibleActionType;
import bluffinmuffin.protocol.TupleTableInfoCareer;
import bluffinmuffin.protocol.TupleTableInfoTraining;
import bluffinmuffin.protocol.commands.lobby.career.CreateCareerTableCommand;
import bluffinmuffin.protocol.commands.lobby.training.CreateTrainingTableCommand;

/**
 * @author Hocus
 *         This class listens to new connection on a certain port number.
 *         This is the main lobby on which client can: connect,
 *         fetch list of available poker tables, create new table and join a table.
 */
public class ServerLobby extends Thread
{
    
    private final int NO_PORT;
    private final ServerSocket m_socketServer;
    private final List<String> m_UsedNames = new ArrayList<String>();
    int m_LastUsedID;
    Map<Integer, PokerGame> m_games = Collections.synchronizedMap(new TreeMap<Integer, PokerGame>());
    
    public PokerGame getGame(int id)
    {
        return m_games.get(id);
    }
    
    public ServerLobby(int p_noPort) throws IOException
    {
        NO_PORT = p_noPort;
        m_socketServer = new ServerSocket(NO_PORT);
    }
    
    public boolean isNameUsed(String name)
    {
        for (final String s : m_UsedNames)
        {
            if (s.equalsIgnoreCase(name))
            {
                return true;
            }
        }
        return false;
    }
    
    public void addName(String name)
    {
        m_UsedNames.add(name);
    }
    
    public void removeName(String name)
    {
        m_UsedNames.remove(name);
    }
    
    public static void main(String args[])
    {
        if ((args.length % 2) == 0)
        {
            try
            {
                final Map<String, String> map = new HashMap<String, String>();
                for (int i = 0; i < args.length; i += 2)
                {
                    map.put(args[i].toLowerCase(), args[i + 1]);
                }
                int port = 4242;
                if (map.containsKey("-p"))
                {
                    port = Integer.parseInt(map.get("-p"));
                }
                final ServerLobby server = new ServerLobby(port);
                server.start();
                System.out.println("Server started on port " + port);
            }
            catch (final Exception e)
            {
                System.err.println("ERROR: Can't start server !!");
            }
        }
        else
        {
            System.err.println("ERROR: incorrect args");
        }
    }
    
    /**
     * Receive all the incoming connection from the clients.
     */
    @Override
    public void run()
    {
        while (true)
        {
            try
            {
                final Socket socketClient = m_socketServer.accept();
                try
                {
                    socketClient.setReuseAddress(true);
                    socketClient.setSoLinger(false, 0);
                    socketClient.setSoTimeout(0);
                }
                catch (final SocketException e1)
                {
                    e1.printStackTrace();
                }
                final ServerClientLobby client = new ServerClientLobby(socketClient, this);
                client.start();
            }
            catch (final IOException e)
            {
                e.printStackTrace();
            }
        }
    }
    
    public int createTrainingTable(CreateTrainingTableCommand command)
    {
        listTrainingTables();
        listCareerTables();
        m_LastUsedID++;
        while (m_games.containsKey(m_LastUsedID))
        {
            m_LastUsedID++;
        }
        final PokerGameTraining game = new PokerGameTraining(new TableInfoTraining(command.getTableName(), command.getBigBlind(), command.getMaxPlayers(), command.getLimit(), command.getStartingMoney()), command.getWaitingTimeAfterPlayerAction(), command.getWaitingTimeAfterBoardDealed(), command.getWaitingTimeAfterPotWon());
        game.start();
        m_games.put(m_LastUsedID, game);
        return m_LastUsedID;
    }
    
    public int createCareerTable(CreateCareerTableCommand command)
    {
        listTrainingTables();
        listCareerTables();
        m_LastUsedID++;
        while (m_games.containsKey(m_LastUsedID))
        {
            m_LastUsedID++;
        }
        final PokerGameCareer game = new PokerGameCareer(new TableInfoCareer(command.getTableName(), command.getBigBlind(), command.getMaxPlayers(), command.getLimit()), command.getWaitingTimeAfterPlayerAction(), command.getWaitingTimeAfterBoardDealed(), command.getWaitingTimeAfterPotWon());
        game.start();
        m_games.put(m_LastUsedID, game);
        return m_LastUsedID;
    }
    
    public synchronized ArrayList<TupleTableInfoCareer> listCareerTables()
    {
        final ArrayList<TupleTableInfoCareer> tables = new ArrayList<TupleTableInfoCareer>();
        final ArrayList<Integer> tablesToRemove = new ArrayList<Integer>();
        
        for (final Integer noPort : m_games.keySet())
        {
            final PokerGame game = m_games.get(noPort);
            
            // Check if the table is still running.
            if (game.isRunning())
            {
                final TableInfo table = game.getTable();
                if (table.getClass().equals(TableInfoCareer.class))
                {
                    tables.add(new TupleTableInfoCareer(noPort, table.getName(), table.getBigBlindAmnt(), table.getPlayers().size(), table.getNbMaxSeats(), table.getBetLimit(), PossibleActionType.None));
                }
            }
            else
            {
                tablesToRemove.add(noPort);
            }
        }
        
        // Remove closed tables.
        for (final Integer key : tablesToRemove)
        {
            m_games.remove(key);
        }
        
        return tables;
    }
    
    public synchronized ArrayList<TupleTableInfoTraining> listTrainingTables()
    {
        final ArrayList<TupleTableInfoTraining> tables = new ArrayList<TupleTableInfoTraining>();
        final ArrayList<Integer> tablesToRemove = new ArrayList<Integer>();
        
        for (final Integer noPort : m_games.keySet())
        {
            final PokerGame game = m_games.get(noPort);
            
            // Check if the table is still running.
            if (game.isRunning())
            {
                final TableInfo table = game.getTable();
                if (table.getClass().equals(TableInfoTraining.class))
                {
                    tables.add(new TupleTableInfoTraining(noPort, table.getName(), table.getBigBlindAmnt(), table.getPlayers().size(), table.getNbMaxSeats(), table.getBetLimit(), PossibleActionType.None));
                }
            }
            else
            {
                tablesToRemove.add(noPort);
            }
        }
        
        // Remove closed tables.
        for (final Integer key : tablesToRemove)
        {
            m_games.remove(key);
        }
        
        return tables;
    }
}
